oci/layout/oci_src_test.go

```diff
@@ -0,0 +1,133 @@
+package layout
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"testing"
+
+	"github.com/containers/image/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const RemoteLayerContent = "This is the remote layer content"
+
+func TestMain(m *testing.M) {
+	httpServer, serverStopSignal, err := startRemoteLayerServer()
+	if err != nil {
+		println("Error starting test TLS server", err.Error())
+		os.Exit(1)
+	}
+
+	code := m.Run()
+
+	stopServer(httpServer, serverStopSignal)
+	os.Exit(code)
+}
+
+func TestGetBlobForRemoteLayers(t *testing.T) {
+	imageSource := createImageSource(t, &types.SystemContext{})
+	layerInfo := types.BlobInfo{
+		URLs: []string{
+			"brokenurl",
+			"https://example.com",
+		},
+	}
+
+	reader, _, err := imageSource.GetBlob(layerInfo)
+	require.NoError(t, err)
+	defer reader.Close()
+
+	data, err := ioutil.ReadAll(reader)
+	require.NoError(t, err)
+	assert.Contains(t, string(data), "Example Domain")
+}
+
+func TestGetBlobForRemoteLayersWithTLS(t *testing.T) {
+	imageSource := createImageSource(t, &types.SystemContext{
+		OCICertPath: "fixtures/accepted_certs",
+	})
+
+	layer, size, err := imageSource.GetBlob(types.BlobInfo{
+		URLs: []string{"https://localhost:12000/"},
+	})
+	require.NoError(t, err)
+
+	layerContent, _ := ioutil.ReadAll(layer)
+	assert.Equal(t, RemoteLayerContent, string(layerContent))
+	assert.Equal(t, int64(len(RemoteLayerContent)), size)
+}
+
+func TestGetBlobForRemoteLayersOnTLSFailure(t *testing.T) {
+	imageSource := createImageSource(t, &types.SystemContext{
+		OCICertPath: "fixtures/rejected_certs",
+	})
+	layer, size, err := imageSource.GetBlob(types.BlobInfo{
+		URLs: []string{"https://localhost:12000/"},
+	})
+
+	require.Error(t, err)
+	assert.Nil(t, layer)
+	assert.Equal(t, int64(0), size)
+}
+
+func remoteLayerContent(w http.ResponseWriter, req *http.Request) {
+	fmt.Fprintf(w, RemoteLayerContent)
+}
+
+func startRemoteLayerServer() (*http.Server, chan struct{}, error) {
+	certBytes, err := ioutil.ReadFile("fixtures/accepted_certs/cert.cert")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	clientCertPool := x509.NewCertPool()
+	if !clientCertPool.AppendCertsFromPEM(certBytes) {
+		return nil, nil, fmt.Errorf("Could not append certificate")
+	}
+
+	tlsConfig := &tls.Config{
+		// Reject any TLS certificate that cannot be validated
+		ClientAuth: tls.RequireAndVerifyClientCert,
+		// Ensure that we only use our "CA" to validate certificates
+		ClientCAs: clientCertPool,
+	}
+
+	tlsConfig.BuildNameToCertificate()
+
+	http.HandleFunc("/", remoteLayerContent)
+
+	httpServer := &http.Server{
+		Addr:      ":12000",
```

Comment:
Nothing guarantees that the port is available. Could this use `:0` to automatically choose an available one?

(There is also `net/http/httptest`; would it simplify things?)

Reply:
Yeah, I think so, I'm going to give it a try here ^^